utils/rpms/daos.spec

```
@@ -351,6 +352,9 @@ mv test.cov{,-build}
       %{?scons_args}                  \
       %{?compiler_args}
 
+# set rpath for libpil4dfs.so is needed to pass daos build test
+patchelf --force-rpath --set-rpath /usr/lib64 %{buildroot}/%{_libdir}/libpil4dfs.so
```

This seems odd.  `/usr/lib64` is a standard search path for the dynamic linker, so why do we have to add it to the `RPATH` here?

Besides that, you should use a pre-defined macro rather than the literal `/usr/lib64`:

```suggestion
patchelf --force-rpath --set-rpath %{_libdir} %{buildroot}/%{_libdir}/libpil4dfs.so
```

But again, it seems redundant to add one of the default search paths to the `RPATH` of something.  Why is this needed?

---

Inside scons daos compiling in daos build test, patchelf is used to set RPATH in the compiled test. Without setting RPATH in libpil4dfs.so compiled in rpm, the preloaded libpil4dfs.so loads some dependent libraries from building directory of daos build test. Setting RPATH in libpil4dfs.so is necessary to pass daos build test in CI.
Thank you! I will replace "/usr/lib64" with "%{_libdir}" in next commit.

---

Can you elaborate on this at all?  As Brian says this is very odd.

---

Maybe I used the daos libraries not from rpm. They contain RPATH pointing to build directory. This causes the issues I previously encountered. I will test and update here later. @brianjmurrell and @ashleypittman Thank you very much!

---

I ran test without setting rpath in libpil4dfs.so at local and in CI. The job in CI is still running. The daos build in local tests always failed. scons sets env "LD_LIBRARY_PATH" as "build/release/gcc/src/gurt:build/release/gcc/src/cart/utils:build/release/gcc/src/cart:/tmp/daos_dfuse_test_dfuse_daos_build_wt_il_1/daos/install/prereq/release/mercury/lib:/tmp/daos_dfuse_test_dfuse_daos_build_wt_il_1/daos/install/prereq/release/isal/lib:build/release/gcc/src/common" in one failed process during compiling DAOS. The libraries from building directories are loaded if rpath is not set in libpil4dfs.so.

$ cat /proc/1776529/maps | grep lib | awk '{print $6}' | sort | uniq
/usr/lib64/libpil4dfs.so
/tmp/daos_dfuse_test_dfuse_daos_build_wt_il_1/daos/build/release/gcc/src/cart/libcart.so.4.9.0
/tmp/daos_dfuse_test_dfuse_daos_build_wt_il_1/daos/build/release/gcc/src/gurt/libgurt.so.4.9.0
/tmp/daos_dfuse_test_dfuse_daos_build_wt_il_1/daos/install/prereq/release/isal/lib/libisal.so.2.0.30
/tmp/daos_dfuse_test_dfuse_daos_build_wt_il_1/daos/install/prereq/release/mercury/lib/libmercury.so.2.3.1
/tmp/daos_dfuse_test_dfuse_daos_build_wt_il_1/daos/install/prereq/release/mercury/lib/libmercury_util.so.4.0.0
/tmp/daos_dfuse_test_dfuse_daos_build_wt_il_1/daos/install/prereq/release/mercury/lib/libna.so.4.0.0
/tmp/daos_dfuse_test_dfuse_daos_build_wt_il_1/daos/install/prereq/release/ofi/lib/libfabric.so.1.22.0
/tmp/daos_dfuse_test_dfuse_daos_build_wt_il_1/daos/install/prereq/release/ucx/lib64/libucm.so.0.0.0
/tmp/daos_dfuse_test_dfuse_daos_build_wt_il_1/daos/install/prereq/release/ucx/lib64/libucp.so.0.0.0
/tmp/daos_dfuse_test_dfuse_daos_build_wt_il_1/daos/install/prereq/release/ucx/lib64/libucs.so.0.0.0
/tmp/daos_dfuse_test_dfuse_daos_build_wt_il_1/daos/install/prereq/release/ucx/lib64/libuct.so.0.0.0
...

---

The test in CI without setting rpath in libpil4dfs.so also failed due to same reason.
gcc: error while loading shared libraries: build/release/gcc/src/client/dfs/libduns.so: file too short

---

@ashleypittman The issue of parallel compiling daos itself with libioil.so in daos build test might be same as the issue here.

---

With setting rpath to /usr/lib64 in libioil.so, parallel compiling daos code itself works without issue with libioil.so in my local test.

---

Parallel compiling daos code itself with libioil.so now works in CI.

---

I was wondering that, great if it is.  Should this be done as part of the build rather than as part of the rpm process however?  @jolivier23 any input here?

It may be worth separating this and the test changes to make it parallel out into it's own PR to make it easier to review/land.

---

Ultimately, setting an `RPATH` to a standard library location is [disallowed in RPM packaged binaries](https://fedoraproject.org/wiki/Changes/Broken_RPATH_will_fail_rpmbuild#Definition_of_a_broken_RPATH).

If I am understanding the problem correctly here, this use of `patchelf` is being used to **undo** (by way of overwriting) an `RPATH` setting being done in the build (and it seems even in an RPM build).

The correct solution, it seems to me is to either not set the `RPATH` in the build in the first place or to **remove** (rather than rewrite/overwrite) the `RPATH` from the binary in the RPM build.  The former would be the more ideal solution.  Really, the `RPATH` should only be getting set by an `scons` build if it can detect that the library is going to be installed into a non-standard location regardless of whether the build is in an RPM context or not.  I.e. even if a user does a build-from-source and the install target library dir is `/usr/lib` or `/usr/lib64`, then no `RPATH` should be set.

Don't we already do this [no-]`RPATH` setting for many other various binaries in the `scons` build process?  I seem to recall this being an issue we resolved very early in the process of building DAOS in RPMs.

---

Setting rpath in libioil.so and libpil4dfs.so was removed in daos.spec file.
Ashley fixed the compiling issue on scons's side in building step. https://github.com/daos-stack/daos/pull/14049

---

> Setting rpath in libioil.so and libpil4dfs.so was removed in daos.spec file. Ashley fixed the compiling issue on scons's side in building step. #14049

Oh, excellent news!  I guess I should have looked at recent commits before adding more comments to this thread.

---

Yes, this problem was tricky but the above change was obviously incorrect for a number of reasons.  The problem was a genuine bug in our build system that had gone unnoticed but is now resolved so this particular part of the PR is a non-issue now.